src/libutil/util.cc

```diff
@@ -786,13 +786,64 @@ pid_t Pid::release()
     return p;
 }
 
-
 void killUser(uid_t uid)
 {
     debug(format("killing all processes running under uid '%1%'") % uid);
 
     assert(uid != 0); /* just to be safe... */
 
+#ifdef __APPLE__
+    /* We originally had very similar logic here as on the Linux side, but
+       then got bitten hard by a system crashing bug as described in here:
+       https://github.com/NixOS/nix/issues/1583. So now we adopt this other
+       approach, but we should go back to the kill approach once Apple fixes
+       their kernel. */
+
+    // Borrowed from https://stackoverflow.com/a/7733928
+    int err = 0;
+    size_t length = 0, newlength = 0;
+
+    // XXX: Should this be KERN_PROC_RUID? EUID vs. real UID
+    static const int name[4] = { CTL_KERN, KERN_PROC, KERN_PROC_UID, static_cast<int>(uid) };
+
+    int tries = 100;
+    bool done = false;
+
+    while (!done) {
+        tries--;
+
+        // First call with NULL to get number of processes. Yeah yeah TOCTOU is fun, but we check
+        // that it hasn't changed later and keep going until it's 0
+        err = sysctl((int *)name, 4, NULL, &length, NULL, 0);
+        if (err)
+            throw SysError("calling sysctl to get size");
+
+        vector<struct kinfo_proc> proc_list(length);
+
+        newlength = length;
+        err = sysctl((int *)name, 4, (void *)proc_list.data(), &newlength, NULL, 0);
+        if (err)
+            throw SysError("calling sysctl to get process list");
+
+        // @grahamc points out https://utcc.utoronto.ca/~cks/space/blog/unix/ProcessKillingTrick
+        for (auto proc : proc_list) {
+            kill(proc.kp_proc.p_pid, SIGSTOP);
+        }
```

Comment:
It is possible that between listing the processes and the SIGSTOP, the process dies and the PID is reused by another user. After the SIGSTOP, we can take a look to see if the PID is owned by the build user we're wanting to kill. If it is our builder, SIGKILL it. If not, SIGCONT it.

@Dezgeg was kind enough to point out SIGCONT to a SIGSTOP'd process, which was SIGSTOP'd before we got to it is rude... which is true, but jeeze, that means the processes exited, the pid recycled, started, and SIGSTOP'd within such a brief period of time. Still rude, but we tried (?)

Reply:
Maybe we should do the `kill`'s as the nixbld user instead?

Reply:
Then if it fails due to EPERM, we just continue on our way

Reply:
🔥

Reply:
@copumpkin Probably this whole loop should just be done as the `nixbld` user (excluding the current process, of course)

Reply:
It'd be really helpful if we could SIGKILL -1 and only kill things we own.

Reply:
Sounds good, I'll run as subprocess again

Reply:
But actually, if I run as subprocess, we need to start worrying about our siblings killing us again...

Reply:
But if our siblings kill us we know there's a rogue